### ISSUE TITLE
Consider an attribute allowing a type to say "this is the CloudEventFormatter I expect to be serialized with"

### ISSUE THREAD

**Issue by jskeet** [cloudevents/sdk-csharp#133]
Assuming https://github.com/googleapis/google-cloudevents-dotnet/pull/59 is merged roughly as-is, the Google.CloudEvents package really just consists of an attribute and some code to create instances of the type referred to by it.

The purpose is to allow a type to say "unless you have other knowledge, this is a good CloudEventFormatter to use when serializing/deserializing CloudEvents that are expected to have this type as data".

So for example, we have:

```csharp
[CloudEventFormatterAttribute(typeof(ProtobufJsonCloudEventFormatter<StorageObjectData>))]
public class StorageObjectData : IMessage<StorageObjectData>
{
   ...
}
```

I'm wondering whether it would make sense to have this attribute in the SDK instead, just as a common way for frameworks to indicate a desired event formatter type.

Assigned to @JoshLove-msft just for the sake of having a think about it before we discuss it further. Other ideas welcome!

**Comment by jskeet** [comment]
(I would say that we probably have more work to do on CloudEventFormatter discovery in general - this is only one piece of the jigsaw. We may well want to make CloudEventFormatters indicate the content types they're capable of handling, for example.)

**Comment by JoshLove-msft** [comment]
Yes, I like this idea. It wouldn't really apply for the Azure CloudEvent as we currently only support JSON but I think it is a nice addition.

**Comment by jskeet** [comment]
We *may* want to add generic classes derived from JsonEventFormatter (both flavours) at the same time, which expect that the data is only ever the JSON representation of the given type - at which point I suspect this could still be useful in Azure as well (as it's not just the format of the data, but the schema that the data represents).

I'll create a PR to discuss at the meeting - probably without tests yet, but that's okay :)
